### ISSUE TITLE
[css-images] Don't force non-legacy colors to interpolate in a gamma-encoded space

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#5883]
Surprisingly, [Color 4](https://www.w3.org/TR/css-color-4/) does not actually have an interpolation section.
[Color 5](https://www.w3.org/TR/css-color-5/) [does](https://drafts.csswg.org/css-color-5/#interpolation), and [Color 4 is even referencing Color 5](https://drafts.csswg.org/css-color-4/#ref-for-angle-value:~:text=This%20impacts%2C%20for%20example%2C%20hue%20interpolation.) in a section.
[Images 3 on the other hand defines its own interpolation rules](https://drafts.csswg.org/css-images-3/#coloring-gradient-line), and specifies that gradients interpolated in premultiplied (gamma-corrected) RGBA, which is clearly from back when all we were dealing with were sRGB colors (though interpolating in gamma-corrected space is suboptimal for those too).

**Where to define interpolation?** Since Color 4 defines the actual color values, Color 4 needs to define how they interpolate. We cannot define values in one level and interpolation in the next level and we can't have level N citing level N+1.

**How to interpolate?** In previous discussions, the rule we had discussed was that [if two colors are in the same color space, they interpolate in that color space](https://drafts.csswg.org/css-color-5/#interpolation-space), otherwise in Lab. After further thinking, I don't think that's a good idea. It does maintain backwards compat for old sRGB forms, but it forces new RGB forms like `color(display-p3)` or even `color(srgb)` (which we resolved in #5825 that is stored separately from legacy sRGB formats) to also interpolate in gamma-corrected space. Instead, I think the rule should be that colors interpolate in Lab (or XYZ), *unless* both colors being interpolated are legacy sRGB forms (for backwards compat) OR the host syntax defines a different interpolation space (e.g. [`color-mix()`](https://drafts.csswg.org/css-color-5/#color-mix)).

**Comment by svgeesus** [comment]
Not that surprising. CSS Color 4 is about specifying colors. (It's also pretty stable and seeing multiple implementations, so will go to CR soon).

CSS Color 5 is about combining and manipulating color - mixing, contrast calculation, and the like. It is therefore natural that it has a section on Color interpolation, which includes [alpha premultiplication](https://drafts.csswg.org/css-color-5/#interpolation-alpha), and [hue interpolation](https://drafts.csswg.org/css-color-5/#hue-interpolation) (that subsection mainly written by @LeaVerou , with corrections by @dbaron).

Color 5 is less mature, though we do have a resolution to publi

So the color interpolation is in the right specification. It is true though that it should probably hace been called something like CSS Color Manipulation 1.

**Comment by svgeesus** [comment]
By the way, as CSSWG already discussed, how to interpolate colors also depends on why you are interpolating _for_. As an example, compositing should be done in a linear-light space like CIE Lab. So Compositing-1 needs updating there; it currently only allows compositing in gamma-encoded sRGB. Filters-1 likewise needs updating (it currently only allows linear-light sRGB). And the definitions of gradients in CSS Images-4 need updating (again they are restricted to gamma-encoded sRGB, so you can't interpolate sRGB to P3 for example).

I'm not prepared to hold up CSS Color 4 while all those other specs get their wider-than-sRGB upgrades.

**Comment by svgeesus** [comment]
>  Instead, I think the rule should be that colors interpolate in Lab (or XYZ), unless both colors being interpolated are legacy sRGB forms (for backwards compat)

That I certainly agree with.

**Comment by dbaron** [comment]
What are you suggesting for handling of alpha in interpolation?

**Comment by svgeesus** [comment]
> OR the host syntax defines a different interpolation space (e.g. color-mix()).

Color-5 [already says that](https://drafts.csswg.org/css-color-5/#interpolation).

> Host syntax can override the interpolation color space and specify which color space is used for interpolation. For example, color-mix and color-adjust override the default to LCH.

**Comment by svgeesus** [comment]
> What are you suggesting for handling of alpha in interpolation?

[Interpolating with alpha](https://drafts.csswg.org/css-color-5/#interpolation-alpha) defines this (for _all_ colorspaces, not just for sRGB.

**Comment by LeaVerou** [comment]
> By the way, as CSSWG already discussed, how to interpolate colors also depends on why you are interpolating for. As an example, compositing should be done in a linear-light space like XYZ. So Compositing-1 needs updating there; it currently only allows compositing in gamma-encoded sRGB. Filters-1 likewise needs updating (it currently only allows linear-light sRGB). And the definitions of gradients in CSS Images-4 need updating (again they are restricted to gamma-encoded sRGB, so you can't interpolate sRGB to P3 for example).
> I'm not prepared to hold up CSS Color 4 while all those other specs get their wider-than-sRGB upgrades.

We should **definitely** not hold up Color 4 until other specs update, but that's not what I was proposing. Color 4 needs to have an interpolation section, which defines interpolation for the common case (gradients, transitions etc), and clarifies that other specs may wish to override, leaving it up to them. We typically define how values interpolate in the same spec where the values are defined. Otherwise, if a UA wishes to implement Color 4 but not Color 5, how are they supposed to implement e.g. gradients between Lab colors? Surely they can't (and shouldn't) follow the Images spec and interpolate them in (s)RGBA.

**Comment by svgeesus** [comment]
> although perhaps some of the text in the examples should also have normative text saying the same thing

Oof, yes, will check for anything only said in examples!

**Comment by svgeesus** [comment]
After further offline discussion with @fantasai and @LeaVerou I plan to move the entire [Color interpolation](https://drafts.csswg.org/css-color-5/#interpolation) section from CSS Color 5 to CSS Color 4.

@tabatkins (for Color 4) @una @argyleink (for Color 5) any objections to that?

The color interpolation section was not in the [First Public Working Draft](https://www.w3.org/TR/2020/WD-css-color-5-20200303/) so there is no patent policy implication. However since Color 4 will be going to CR soon, I want to get it into Color 4 before that, so that we get the patent policy protection on that section as soon as the CR exclusion period ends.

**Comment by argyleink** [comment]
no objections 👍🏻

**Comment by svgeesus** [comment]
@LeaVerou  wrote:

> Where to define interpolation? Since Color 4 defines the actual color values, Color 4 needs to define how they interpolate. We cannot define values in one level and interpolation in the next level and we can't have level N citing level N+1.

[done](https://github.com/w3c/csswg-drafts/commit/09bc009d01321f3581b901c1a11f1b2f51fa8913)

> How to interpolate? In previous discussions, the rule we had discussed was that if two colors are in the same color space, they interpolate in that color space, otherwise in Lab. After further thinking, I don't think that's a good idea. It does maintain backwards compat for old sRGB forms, but it forces new RGB forms like color(display-p3) or even color(srgb) (which we resolved in #5825 that is stored separately from legacy sRGB formats) to also interpolate in gamma-corrected space. Instead, I think the rule should be that colors interpolate in Lab (or XYZ), unless both colors being interpolated are legacy sRGB forms (for backwards compat) OR the host syntax defines a different interpolation space (e.g. color-mix()).

[also done](https://github.com/w3c/csswg-drafts/commit/aba4ebb74a761892d298f85905d11bd4eaea9a7c)

Still needs some examples, but please review.

**Comment by Myndex** [comment]
> ....compositing should be done in a linear-light space like CIE Lab....

I do agree that *most* compositing is *usually* best in a linearized space.

BUT..... LAB is not a linear light space, and I know you know LAB and LUV are perceptual, so not sure which space was actually meant?

In Film/TV **LUV** or a variant is commonly used, at least "in the guts" of color grading gear. **LAB** is better suited to reflected colors. I don't remember any time where I used either for compositing though, only for adjusting/grading.

In film/TV typically we linearize RGB. In particular, when possible we linearize the destination or delivery workspace, then the only transform needed on output is applying the transfer curve.

Here's a GIST comparing LAB and LUV:  

[**Where's the LUV?**](https://gist.github.com/Myndex/47c793f8a054041bd2b52caa7ad5271c)

**Comment by svgeesus** [comment]
@Myndex wrote:

> In Film/TV LUV or a variant is commonly used, at least "in the guts" of color grading gear. LAB is better suited to reflected colors.

I have seen this "LUV for emissive color, LAB for reflective color" meme repeatedly, but never with any actual evidence. I believe it can be traced back to the 1976 CIE committee that standardized both LUV and LAB, and was given as a rather handwavy justification for standardizing both.

Since then, ever paper or book or system that I have seen uses LAB and not LUV, unless:

- it has moved on from colorimetry to color appearance modelling and thus uses CIECAM16 or whatever
- it has moved on to HDR and thus (apart from Fairchild's HDR Lab) it uses IVCtCp, or Jzazbz
- it is primarily concerned with modelling physical light transport, in which case it uses a linear-light model such as XYZ (or any linearized RGB space, provided it can handle out-of-range colors correctly)
- it is illustrating gamuts in a vague and handwavy way, and thus is satisfied with a two-dimensional gamut representation, glossing over the fact that out of gamut colors can be inside the gamut triangle. It therefore uses the u'v' chromaticity diagram, which is not very perceptually uniform but is less broken than the xy one. More rigorous work is done with a 3D gamut volume, typically in Lab or in UCS-16 Jab.

Checking just now the indexes for six books on various color science topics, LAB is mentioned all over and LUV gets a single mention near the beginning, mostly to note that it is not used, has poor distance metrics, and gives wildly inaccurate predictions for corresponding colors.

If you have examples of CIE LUV being used, please cite them.

If you have examples of a system derived from LUV being used, please name that system.

**Comment by svgeesus** [comment]
> BUT..... LAB is not a linear light space, and I know you know LAB and LUV are perceptual, so not sure which space was actually meant?

I meant to type CIE XYZ. Comment above edited to correct it. Thanks for noticing!

**Comment by Myndex** [comment]
Hey Chris, @svgeesus 

> _I have seen this "LUV for emissive color, LAB for reflective color" meme repeatedly, but never with any actual evidence._

Yea, I've seen that, and had not given it too much weight until I began some comparative evaluations — I hope you know I question everything and believe no one! (I come from the "question authority" generation after all, LOL). 😎

> _and was given as a rather handwavy justification for standardizing both._

And, both LAB and LUV are a little handwavy. But also computationally simple compared to complete models.

> _ever paper or book or system that I have seen uses LAB and not LUV,_

I do rely a bit on R.W.G. Hunt's **The Reproduction of Colour** where he says:

_"This quantity is called the CIE 1976 (L*u*v*) colour difference ....This type of space is useful in studies of colour television because, in the associated chromaticity diagrams, additive mixtures of stimuli lie on straight lines joining the points representing those stimuli..."_

Otherwise YES I do agree, especially for general color appearance modeling. LUV is near the bottom of the bunch for general purpose appearance modeling and reflective colors, and therefore not given much weight in many books on the subject. In particular I know Mark Fairchild dismisses LUV pretty harshly, and indeed LUV is NOT useful for color differences of reflected/surface colors.

Nevertheless, it seems somewhat prominent in _" HIGH DYNAMIC RANGE IMAGING (2010)"_ though that is probably due to SGI's development of LogLuv.

And stepping away from pure color theory, data visualization has an affinity for LUV over LAB. From "INFORMATION
VISUALIZATION Perception for Design 2nd ed":

_"...These are called the CIElab and the CIEluv uniform color spaces.....the two standards have somewhat different properties that make them useful for different tasks. **Only the CIEluv formula is described here.** It is generally held to be better for specifying large color differences...."_

> _If you have examples of CIE LUV being used, please cite them. If you have examples of a system derived from LUV being used, please name that system._

I _have_ cited/mentioned/examples....  I realize I write a lot, not sure if you have ever received any of the direct emails I've sent you? I also discuss this more in depth in issue #6061 and below are links to 

I have a Gist with examples, links, etc: which was linked in the above message, again:  [**Where's The Luv?**](https://gist.github.com/Myndex/47c793f8a054041bd2b52caa7ad5271c)

### Some Use Cases
- MilSpec specifies uv coordinates for lighting (i.e. such as for signaling colors to be used in dark adapted environments such as a cockpit).
- Tektronix color grading systems use LUV (with a minor mod), as shown in their patents.
- SGI created LogLuv for HDR
- CalMAN™ uses CIELuv
- Data Vis: https://ccom.unh.edu/vislab/projects/ColorSequenceEditor.html
- Data Vis: https://www.csc2.ncsu.edu/faculty/healey/download/sig-course.98.pdf

I've also noticed as I started digging deeper into neurological research the use of CIELUV for research work, such as in this paper: https://www.nature.com/articles/srep42364.pdf

**What Caused my Interest:**
As it happens, in developing SAPC I was working more with LAB — but being often frustrated with a lack of fit to collected data, I started working more with LUV, and also examining other models. TO BE CLEAR: I am not trying to dismiss LAB in any way, but I did add LUV to the models I'm working with, and I've found several aspects that make it favorable over LAB on self illuminated displays.

- I AGREE that Jzazbz is better... it is also more complicated and therefore computationally more expensive.
- Luv is a simple model, 
    - only a single exponent for `L*` which is identical to Lab.
    - The uv is just a scaling of the 1976 UCS.
    - Simplification: Luv is just xyY with Y transformed to `L*` and then `L*` being used as part of scaling the xy
    - As such I *think* I could create a matrix that goes direct from sRGB into Luv without a transit through XYZ (don't hold me to that, I have not fully investigated).
- Regarding usefulness in mixing and gradients, I am also still evaluating (read: eliminating confirmation biases by waiting weeks and reviewing with fresh eyes).
- I have been working on programatic perceptual color sorting... Luv with an eccentric hue based modifier I'm experimenting with "seems" to give useful (better than LAB) results.... 
- Additional examples in the form of artist colorwheels in this brief article I wrote a while ago looks at Lab Luv and others, my conclusion at the time was that neither Lab nor Luv created accurate color wheels, but Luv was better — nevertheless, manual (LUT based) spacing was needed. Article:  https://www.myndex.com/WDEV/colorwheel

## TL;DR

All that said, I am not the publicist for LUV! LOL ... just pointing out that it has useful applications, and my own experiments and comparisons have shown clear advantages over LAB for use on self illuminated displays for color picking.


Thank you!!

Andy

**Comment by svgeesus** [comment]
> I do rely a bit on R.W.G. Hunt's The Reproduction of Colour where he says:

Hunts' work is well regarded and was seminal in it's day, but the first edition dates to 1957. The fifth edition, which I have, is from 1995 and is a somewhat light re-warming of the 1987 fourth edition. So in terms of evaluating the impact of the 1976 color models, it necessarily misses a lot of later work.

> "This quantity is called the CIE 1976 (Luv*) colour difference 

Yes, it has a color difference formula. Comparative studies tend to rate it last, and more modern studies don't even evaluate it at all.

> This type of space is useful in studies of colour television because, in the associated chromaticity diagrams

I mentioned the (limited) value of chromaticity diagrams as one use case. For stage lighting design, in particular, they were fairly useful This was in the days before personal computers or even programmable calculators, it was a benefit to plot light sources on a large chromaticity diagram drawn on graph paper, connect them with a straight line, and be able to directly read off the resulting mixture chromaticities to two or perhaps three decimal places.

Nowadays they are typically used in reviews or monitors or TVs, as a first approximation to show the gamut of one device compared to another. But due to the lack of perceptual uniformity it is easy to be led astray here. The MacAdam ellipses are still very elliptical and vary greatly in size.

> LUV is near the bottom of the bunch for general purpose appearance modeling and reflective colors,

Also for emissive colors. The flaws of the model are not suddenly cured because the light from a colored patch is directly generated rather than being the modified, reflected light. And there is only a small difference in geometry separating an opaque patch illuminated from above, and a translucent filter illuminated from behind by a backlight, as in LCD displays.

> " HIGH DYNAMIC RANGE IMAGING (2010)"

(I don't have that one yet, it is eye-wateringly expensive and I fund all my book purchases; W3C doesn't cover them.)

> CalMAN™ uses CIELuv

No, CalMan used to use Lab, with deltaE2000 for all color differences, and now has [moved to using ICtCp, with deltaITP](https://kb.portrait.com/help/ictcp-color-difference-metric).

>  I am not trying to dismiss LAB in any way, but I did add LUV to the models I'm working with, and I've found several aspects that make it favorable over LAB on self illuminated displays.

Thanks for the clarification. Some of your comments, and in particular the personal mails you sent to me earlier, gave a very different impression where you seemed to suggest that any use of Lab was entirely erroneous and purely due to commercial pressure from one or two companies. I'm glad to hear you state more clearly your position - thanks!

>    I AGREE that Jzazbz is better.

It is better in some ways. I find that the dreaded blue-to-purple shift in Lab is a little over-compensated, giving a turquoise case to blue chroma reduction. It also depends more on the state of adaptation and is sensitive to the scaling (where media white is placed, on the absolute luminance scale used in PQ).

Lastly, while this is interesting and useful discussion, it is entirely unrelated to the topic of this particular issue "Don't force non-legacy colors to interpolate in a gamma-encoded space".

I do hope, though, that your more wide-ranging views will be brought to bear during the [Color Workshop](https://www.w3.org/Graphics/Color/Workshop/). I would also encourage you to submit your research to a Color Science journal.

**Comment by Myndex** [comment]
Hi @svgeesus  I moved my reply to the approriate thread:

https://github.com/w3c/csswg-drafts/issues/6061#issuecomment-790304661

**Comment by svgeesus** [comment]
To recap, the color interpolation section has been moved from Color 5 to Color 4 and the requested edits have been made to that section.

The remaining part is

> Images 3 on the other hand defines its own interpolation rules, and specifies that gradients interpolated in premultiplied (gamma-corrected) RGBA, which is clearly from back when all we were dealing with were sRGB colors (though interpolating in gamma-corrected space is suboptimal for those too).

Re-tagging therefore to Images 4, which is the current work.

**Comment by Smilebags** [comment]
I would like to read up on any recent progress on Images 4, in particular whether there is anything regarding the space in which compositing occurs (essentially the equivalent of the new CSS colour interpolation options). I am curious to see the direction in which the web is going in terms of how WCG and/or HDR content will be composited in a webpage. Is there any draft specifications in this area?

**Comment by svgeesus** [comment]
@Smilebags interpolation for gradients would be defined in images 4 (with perceptually uniform spaced like Lab and CH being the most desirable options), but interpolation for compositing would be defined in CSS Compositing; it currently has a gamma-encoded sRGB default and an update would add CIE XYZ because there needs to be a linear-light option.

**Comment by Smilebags** [comment]
Thanks for that information @svgeesus.
I have been keeping an eye on OKLab and was excited to see your recent presentation on it. I'm optimistic that it will provide a very nice modern default for gradient and colour scheme generation, in particular in domains such as data visualisation and design systems.
I agree the most problematic part of compositing on the web is the lack of an option for linear-light blending. That seems like it should also simplify (or at least be a part of) the solution to compositing of HDR and WCG content into a webpage.

**Comment by svgeesus** [comment]
There were some good presentations about HDR and HDR/SDR compositing at the workshop:

[HDR: Compositing and tone mapping](https://www.w3.org/Graphics/Color/Workshop/talks.html#compos)

**Comment by LeaVerou** [comment]
> To recap, the color interpolation section has been moved from Color 5 to Color 4 and the requested edits have been made to that section.
> 
> The remaining part is
> 
> > Images 3 on the other hand defines its own interpolation rules, and specifies that gradients interpolated in premultiplied (gamma-corrected) RGBA, which is clearly from back when all we were dealing with were sRGB colors (though interpolating in gamma-corrected space is suboptimal for those too).
> 
> Re-tagging therefore to Images 4, which is the current work.

Since color interpolation syntax was just added to gradient functions in https://github.com/w3c/csswg-drafts/commit/1f940c6aa92606b447b9b10f880d089833511f77 , I’m gonna go ahead and close this.
